### PR TITLE
CompanyService_AddCompany

### PR DESCRIPTION
**Pull request by KostinP**

### CODE REVIEW

File: DigitalOffice/Services/CompanyService/Controllers/CompanyController.cs [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+using System;
+
+namespace LT.DigitalOffice.CompanyService.Controllers
+{
+    [Route("api/[controller]")]
+    [ApiController]
+    public class CompanyController : ControllerBase
+    {
+        [HttpPost("addCompany")]
+        public Guid CreateUser([FromServices] IAddCompanyCommand command, [FromBody] AddCompanyRequest request)
```
Comment (lalagvana) [inline DigitalOffice/Services/CompanyService/Controllers/CompanyController.cs]: Why CreateUser?

File: DigitalOffice/Tests/CompanyServiceTests/Validators/AddCompanyValidatorTests.cs [hunk 2]
```diff
@@ -0,0 +1,40 @@
+﻿using System;
```
Comment (MihailKulikov) [inline DigitalOffice/Tests/CompanyServiceTests/Validators/AddCompanyValidatorTests.cs]: Using directive is not required by the code and can be safely removed

**Review by MihailKulikov — APPROVED** [review]

File: DigitalOffice/Services/CompanyService/Commands/AddCompanyCommand.cs [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+            this.validator = validator;
+            this.repository = repository;
+            this.mapper = mapper;
+        }
+
+        public Guid Execute(AddCompanyRequest request)
+        {
+            validator.ValidateAndThrow(request);
+
+            var user = mapper.Map(request);
```
Comment (MihailKulikov) [inline DigitalOffice/Services/CompanyService/Commands/AddCompanyCommand.cs]: user?

File: DigitalOffice/Tests/CompanyServiceTests/Commands/AddCompanyCommandTests.cs [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+            validatorMock = new Mock<IValidator<AddCompanyRequest>>();
+            mapperMock = new Mock<IMapper<AddCompanyRequest, DbCompany>>();
+            repositoryMock = new Mock<ICompanyRepository>();
+
+            command = new AddCompanyCommand(validatorMock.Object, mapperMock.Object, repositoryMock.Object);
+        }
+
+
+        [Test]
+        public void FailCreateNewWorkTimeAccordingToValidator()
```
Comment (MihailKulikov) [inline DigitalOffice/Tests/CompanyServiceTests/Commands/AddCompanyCommandTests.cs]: Work time?

File: DigitalOffice/Tests/CompanyServiceTests/Commands/AddCompanyCommandTests.cs [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+            validatorMock
+                .Setup(x => x.Validate(It.IsAny<IValidationContext>()).IsValid)
+                .Returns(false);
+
+            Assert.Throws<ValidationException>(() => command.Execute(request));
+            repositoryMock.Verify(repository => repository.CreateCompany(It.IsAny<DbCompany>()), Times.Never);
+        }
+
+        [Test]
+        public void FailCreateNewWorkTimeAccordingToRepository()
```
Comment (MihailKulikov) [inline DigitalOffice/Tests/CompanyServiceTests/Commands/AddCompanyCommandTests.cs]: Work time?

File: DigitalOffice/Tests/CompanyServiceTests/Commands/AddCompanyCommandTests.cs [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+
+            repositoryMock
+                .Setup(x => x.CreateCompany(It.IsAny<DbCompany>()))
+                .Throws(new Exception());
+
+            Assert.Throws<Exception>(() => command.Execute(request));
+        }
+
+        [Test]
+        public void SuccessfulCreateNewWorkTime()
```
Comment (MihailKulikov) [inline DigitalOffice/Tests/CompanyServiceTests/Commands/AddCompanyCommandTests.cs]: Work time?

**Review by MihailKulikov — CHANGES_REQUESTED** [review]

File: DigitalOffice/Tests/CompanyServiceTests/Commands/AddCompanyCommandTests.cs (hunk 6 above)
Comment (KostinP) [reply DigitalOffice/Tests/CompanyServiceTests/Commands/AddCompanyCommandTests.cs]: Fixed.

File: DigitalOffice/Tests/CompanyServiceTests/Commands/AddCompanyCommandTests.cs (hunk 5 above)
Comment (KostinP) [reply DigitalOffice/Tests/CompanyServiceTests/Commands/AddCompanyCommandTests.cs]: Fixed.

File: DigitalOffice/Tests/CompanyServiceTests/Commands/AddCompanyCommandTests.cs (hunk 4 above)
Comment (KostinP) [reply DigitalOffice/Tests/CompanyServiceTests/Commands/AddCompanyCommandTests.cs]: Fixed.

File: DigitalOffice/Services/CompanyService/Commands/AddCompanyCommand.cs (hunk 3 above)
Comment (KostinP) [reply DigitalOffice/Services/CompanyService/Commands/AddCompanyCommand.cs]: Fixed.

File: DigitalOffice/Tests/CompanyServiceTests/Validators/AddCompanyValidatorTests.cs (hunk 2 above)
Comment (KostinP) [reply DigitalOffice/Tests/CompanyServiceTests/Validators/AddCompanyValidatorTests.cs]: Fixed.

File: DigitalOffice/Services/CompanyService/Controllers/CompanyController.cs (hunk 1 above)
Comment (KostinP) [reply DigitalOffice/Services/CompanyService/Controllers/CompanyController.cs]: Fixed.

**Review by lalagvana — APPROVED** [review]

**Review by mekk1t — APPROVED** [review]

**Review by MihailKulikov — APPROVED** [review]

**Review by ifelseelif — APPROVED** [review]

File: DigitalOffice/Services/CompanyService/Validators/AddCompanyValidator.cs [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+
+namespace LT.DigitalOffice.CompanyService.Validators
+{
+    public class AddCompanyValidator : AbstractValidator<AddCompanyRequest>
+    {
+        public AddCompanyValidator()
+        {
+            RuleFor(c => c.Name)
+                .NotEmpty();
+        }
```
Comment (denis-kriv) [inline DigitalOffice/Services/CompanyService/Validators/AddCompanyValidator.cs]: how about rules for length of name?

Comment (KostinP) [reply DigitalOffice/Services/CompanyService/Validators/AddCompanyValidator.cs]: Fixed.

File: DigitalOffice/Tests/CompanyServiceTests/Commands/AddCompanyCommandTests.cs [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+            validatorMock = new Mock<IValidator<AddCompanyRequest>>();
+            mapperMock = new Mock<IMapper<AddCompanyRequest, DbCompany>>();
+            repositoryMock = new Mock<ICompanyRepository>();
+
+            command = new AddCompanyCommand(validatorMock.Object, mapperMock.Object, repositoryMock.Object);
+        }
+
+
+        [Test]
+        public void FailAddCompanyAccordingToValidator()
```
Comment (lalagvana) [inline DigitalOffice/Tests/CompanyServiceTests/Commands/AddCompanyCommandTests.cs]: Rename tests as we decided on our last conversation. With "should"

File: DigitalOffice/Tests/CompanyServiceTests/Mappers/CompanyMapperTests.cs [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+        }
+
+        [Test]
+        public void ThrowExceptionIfArgumentIsNull()
+        {
+            Assert.Throws<ArgumentNullException>(() => mapper.Map(null));
+        }
+
+        [Test]
+        public void ReturnRightModelSuccessfully()
```
Comment (lalagvana) [inline DigitalOffice/Tests/CompanyServiceTests/Mappers/CompanyMapperTests.cs]: Hm, strange name

File: DigitalOffice/Tests/CompanyServiceTests/Mappers/CompanyMapperTests.cs [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+        {
+            var newWorkTime = mapper.Map(request);
```
Comment (lalagvana) [inline DigitalOffice/Tests/CompanyServiceTests/Mappers/CompanyMapperTests.cs]: WorkTime?

File: DigitalOffice/Tests/CompanyServiceTests/Commands/AddCompanyCommandTests.cs (hunk 8 above)
Comment (KostinP) [reply DigitalOffice/Tests/CompanyServiceTests/Commands/AddCompanyCommandTests.cs]: Fixed.

File: DigitalOffice/Tests/CompanyServiceTests/Mappers/CompanyMapperTests.cs (hunk 9 above)
Comment (KostinP) [reply DigitalOffice/Tests/CompanyServiceTests/Mappers/CompanyMapperTests.cs]: Fixed.

File: DigitalOffice/Tests/CompanyServiceTests/Mappers/CompanyMapperTests.cs (hunk 10 above)
Comment (KostinP) [reply DigitalOffice/Tests/CompanyServiceTests/Mappers/CompanyMapperTests.cs]: Fixed.

**Review by djino97 — APPROVED** [review]

**Pull request merged**